File: native/src/struct_features.cpp
#include "clang_ldl/struct_features.hpp"

#include <algorithm>
#include <cmath>
#include <vector>

namespace clang_ldl {

namespace {

int holes_count(const uint8_t* bitmap, int width, int height) {
    const int pw = width + 2;
    const int ph = height + 2;
    std::vector<uint8_t> grid(static_cast<size_t>(pw) * ph, 0);
    for (int y = 0; y < height; ++y) {
        for (int x = 0; x < width; ++x) {
            if (bitmap[static_cast<size_t>(y) * width + x] > 127) {
                grid[static_cast<size_t>(y + 1) * pw + (x + 1)] = 1;
            }
        }
    }
    std::vector<uint8_t> visited(grid.size(), 0);
    int holes = 0;
    auto idx = [pw](int x, int y) { return static_cast<size_t>(y) * pw + x; };

    for (int y = 0; y < ph; ++y) {
        for (int x = 0; x < pw; ++x) {
            if (grid[idx(x, y)] != 0 || visited[idx(x, y)]) {
                continue;
            }
            bool touches_border = false;
            std::vector<std::pair<int, int>> stack = {{x, y}};
            while (!stack.empty()) {
                auto [cx, cy] = stack.back();
                stack.pop_back();
                const size_t i = idx(cx, cy);
                if (visited[i] || grid[i] != 0) {
                    continue;
                }
                visited[i] = 1;
                if (cx == 0 || cy == 0 || cx == pw - 1 || cy == ph - 1) {
                    touches_border = true;
                }
                const int dx[4] = {1, -1, 0, 0};
                const int dy[4] = {0, 0, 1, -1};
                for (int k = 0; k < 4; ++k) {
                    const int nx = cx + dx[k];
                    const int ny = cy + dy[k];
                    if (nx >= 0 && ny >= 0 && nx < pw && ny < ph) {
                        stack.emplace_back(nx, ny);
                    }
                }
            }
            if (!touches_border) {
                ++holes;
            }
        }
    }
    return holes;
}

void skeleton_pass(std::vector<uint8_t>& src, int width, int height, bool endpoints) {
    std::vector<uint8_t> dst = src;
    for (int y = 1; y < height - 1; ++y) {
        for (int x = 1; x < width - 1; ++x) {
            if (src[static_cast<size_t>(y) * width + x] <= 127) {
                continue;
            }
            int n = 0;
            for (int dy = -1; dy <= 1; ++dy) {
                for (int dx = -1; dx <= 1; ++dx) {
                    if (dx == 0 && dy == 0) {
                        continue;
                    }
                    if (src[static_cast<size_t>(y + dy) * width + (x + dx)] > 127) {
                        ++n;
                    }
                }
            }
            if (endpoints ? n < 2 : n == 4) {
                dst[static_cast<size_t>(y) * width + x] = 0;
            }
        }
    }
    src.swap(dst);
}

void skeleton_stats(const uint8_t* bitmap, int width, int height, uint8_t& endpoints, uint8_t& junctions) {
    std::vector<uint8_t> skel(bitmap, bitmap + static_cast<size_t>(width) * height);
    bool changed = true;
    int iter = 0;
    while (changed && iter < 32) {
        changed = false;
        const auto before = skel;
        skeleton_pass(skel, width, height, true);
        skeleton_pass(skel, width, height, false);
        if (skel != before) {
            changed = true;
        }
        ++iter;
    }
    int ep = 0;
    int jn = 0;
    for (int y = 0; y < height; ++y) {
        for (int x = 0; x < width; ++x) {
            if (skel[static_cast<size_t>(y) * width + x] <= 127) {
                continue;
            }
            int n = 0;
            for (int dy = -1; dy <= 1; ++dy) {
                for (int dx = -1; dx <= 1; ++dx) {
                    if (dx == 0 && dy == 0) {
                        continue;
                    }
                    const int nx = x + dx;
                    const int ny = y + dy;
                    if (nx < 0 || ny < 0 || nx >= width || ny >= height) {
                        continue;
                    }
                    if (skel[static_cast<size_t>(ny) * width + nx] > 127) {
                        ++n;
                    }
                }
            }
            if (n <= 1) {
                ++ep;
            } else if (n >= 3) {
                ++jn;
            }
        }
    }
    endpoints = static_cast<uint8_t>(std::min(ep, 255));
    junctions = static_cast<uint8_t>(std::min(jn, 255));
}

} // namespace

StructFeatures compute_struct_features(const uint8_t* bitmap, int width, int height) {
    StructFeatures f;
    if (!bitmap || width <= 0 || height <= 0) {
        return f;
    }
    int minx = width, miny = height, maxx = 0, maxy = 0;
    int ink = 0;
    for (int y = 0; y < height; ++y) {
        for (int x = 0; x < width; ++x) {
            if (bitmap[static_cast<size_t>(y) * width + x] > 127) {
                ++ink;
                minx = std::min(minx, x);
                maxx = std::max(maxx, x);
                miny = std::min(miny, y);
                maxy = std::max(maxy, y);
            }
        }
    }
    if (ink == 0) {
        return f;
    }
    const int bw = maxx - minx + 1;
    const int bh = maxy - miny + 1;
    f.aspect = static_cast<float>(bw) / static_cast<float>(std::max(bh, 1));
    f.holes = static_cast<uint8_t>(std::min(holes_count(bitmap, width, height), 255));
    skeleton_stats(bitmap, width, height, f.endpoints, f.junctions);
    return f;
}

} // namespace clang_ldl

File: tools/struct_features.py
"""Structural features for glyph bitmaps (holes, skeleton stats)."""

from __future__ import annotations


def count_holes(bitmap: list[int], width: int, height: int) -> int:
    """Count enclosed background regions (holes) in a binary ink mask."""
    padded_w = width + 2
    padded_h = height + 2
    grid = [0] * (padded_w * padded_h)
    for y in range(height):
        for x in range(width):
            if bitmap[y * width + x] > 127:
                grid[(y + 1) * padded_w + (x + 1)] = 1

    visited = [False] * (padded_w * padded_h)
    holes = 0

    def idx(x: int, y: int) -> int:
        return y * padded_w + x

    def flood(x: int, y: int) -> bool:
        stack = [(x, y)]
        touches_border = False
        while stack:
            cx, cy = stack.pop()
            i = idx(cx, cy)
            if visited[i] or grid[i] != 0:
                continue
            visited[i] = True
            if cx == 0 or cy == 0 or cx == padded_w - 1 or cy == padded_h - 1:
                touches_border = True
            for dx, dy in ((1, 0), (-1, 0), (0, 1), (0, -1)):
                nx, ny = cx + dx, cy + dy
                if 0 <= nx < padded_w and 0 <= ny < padded_h:
                    stack.append((nx, ny))
        return not touches_border

    for y in range(padded_h):
        for x in range(padded_w):
            if grid[idx(x, y)] == 0 and not visited[idx(x, y)]:
                if flood(x, y):
                    holes += 1
    return holes


def _skeletonize(bitmap: list[int], width: int, height: int) -> list[int]:
    """Simple morphological skeleton approximation."""
    src = bitmap[:]
    dst = [0] * len(src)
    changed = True
    while changed:
        changed = False
        dst = src[:]
        for y in range(1, height - 1):
            for x in range(1, width - 1):
                if src[y * width + x] <= 127:
                    continue
                neighbors = sum(
                    1
                    for dx, dy in ((-1, 0), (1, 0), (0, -1), (0, 1), (-1, -1), (1, -1), (-1, 1), (1, 1))
                    if src[(y + dy) * width + (x + dx)] > 127
                )
                if neighbors < 2:
                    dst[y * width + x] = 0
                    changed = True
        src = dst[:]
        for y in range(1, height - 1):
            for x in range(1, width - 1):
                if src[y * width + x] <= 127:
                    continue
                neighbors = sum(
                    1
                    for dx, dy in ((-1, 0), (1, 0), (0, -1), (0, 1))
                    if src[(y + dy) * width + (x + dx)] > 127
                )
                if neighbors == 4:
                    dst[y * width + x] = 0
                    changed = True
        src = dst[:]
    return src


def skeleton_stats(bitmap: list[int], width: int, height: int) -> tuple[int, int]:
    skel = _skeletonize(bitmap, width, height)
    endpoints = 0
    junctions = 0
    for y in range(height):
        for x in range(width):
            if skel[y * width + x] <= 127:
                continue
            n = sum(
                1
                for dx, dy in ((-1, 0), (1, 0), (0, -1), (0, 1), (-1, -1), (1, -1), (-1, 1), (1, 1))
                if 0 <= x + dx < width
                and 0 <= y + dy < height
                and skel[(y + dy) * width + (x + dx)] > 127
            )
            if n <= 1:
                endpoints += 1
            elif n >= 3:
                junctions += 1
    return endpoints, junctions


def compute_features(bitmap: list[int], width: int, height: int) -> dict[str, float | int]:
    ink = sum(1 for v in bitmap if v > 127)
    ys = [y for y in range(height) for x in range(width) if bitmap[y * width + x] > 127]
    xs = [x for y in range(height) for x in range(width) if bitmap[y * width + x] > 127]
    if ink == 0:
        return {"holes": 0, "endpoints": 0, "junctions": 0, "aspect": 1.0}
    bw = max(xs) - min(xs) + 1
    bh = max(ys) - min(ys) + 1
    holes = count_holes(bitmap, width, height)
    endpoints, junctions = skeleton_stats(bitmap, width, height)
    return {
        "holes": holes,
        "endpoints": min(endpoints, 255),
        "junctions": min(junctions, 255),
        "aspect": bw / max(bh, 1),
    }
